example-app/src/main/java/com/adyen/checkout/example/service/ExampleAsyncDropInService.kt

```diff
@@ -98,29 +105,72 @@ class ExampleAsyncDropInService : DropInService() {
-        return if (response != null) {
-            val detailsResponse = JSONObject(response.string())
-            if (detailsResponse.has("action")) {
-                DropInServiceResult.Action(detailsResponse.get("action").toString())
-            } else {
-                Logger.d(TAG, "Final result - ${detailsResponse.toStringPretty()}")
-
-                val resultCode = if (detailsResponse.has("resultCode")) {
-                    detailsResponse.get("resultCode").toString()
+    private fun handleResponse(response: ResponseBody?): DropInServiceResult? {
+        val jsonResponse = if (response == null) null else JSONObject(response.string())
+        return when {
+            jsonResponse == null -> {
+                Logger.e(TAG, "FAILED")
+                DropInServiceResult.Error(reason = "IOException")
+            }
+            isAction(jsonResponse) -> {
+                Logger.d(TAG, "Received action")
+                val action = Action.SERIALIZER.deserialize(jsonResponse.getJSONObject("action"))
+                DropInServiceResult.Action(action)
+            }
+            isNonFullyPaidOrder(jsonResponse) -> {
+                Logger.d(TAG, "Received a non fully paid order")
+                fetchPaymentMethods(jsonResponse)
+                null
+            }
+            else -> {
+                Logger.d(TAG, "Final result - ${jsonResponse.toStringPretty()}")
+                val resultCode = if (jsonResponse.has("resultCode")) {
+                    jsonResponse.get("resultCode").toString()
                 } else {
                     "EMPTY"
                 }
                 DropInServiceResult.Finished(resultCode)
             }
-        } else {
-            Logger.e(TAG, "FAILED")
-            DropInServiceResult.Error(reason = "IOException")
+        }
+    }
+
+    private fun isAction(jsonResponse: JSONObject): Boolean {
+        return jsonResponse.has("action")
+    }
+
+    private fun isNonFullyPaidOrder(jsonResponse: JSONObject): Boolean {
+        return jsonResponse.has("order") && getOrderFromResponse(jsonResponse).remainingAmount?.value ?: 0 > 0
+    }
+
+    private fun getOrderFromResponse(jsonResponse: JSONObject): OrderResponse {
+        val orderJSON = jsonResponse.getJSONObject("order")
+        return OrderResponse.SERIALIZER.deserialize(orderJSON)
+    }
+
+    private fun fetchPaymentMethods(jsonResponse: JSONObject) {
+        launch(Dispatchers.IO) {
+            val order = getOrderFromResponse(jsonResponse)
+            val paymentMethods = paymentsRepository.getPaymentMethods(
+                getPaymentMethodRequest(
+                    keyValueStorage,
+                    Order(
+                        pspReference = order.pspReference,
+                        orderData = order.orderData
+                    )
+                )
+            )
+            val result = if (paymentMethods != null) {
+                DropInServiceResult.Update(paymentMethods, order)
```

Comment:
Why is this in the Example class? I thought we would be the ones calling the updated `paymentMethods` response after the Order no?

Because this is triggering the DropInActivity to change the Fragment back to the Payment Methods list right?

Reply:
I thought so too but apparently it's all the merchant's responsibility. They need to check if the response of `/payments` contains an `order` which is not fully paid and then call `paymentMethods` again while passing the `order` in the request, then give us the list back. Same goes for cancelling an order.